cryoet_data_portal_neuroglancer/precompute/contrast_limits.py

```diff
@@ -0,0 +1,557 @@
+            try:
+                gmm.fit(sample_data.reshape(-1, 1))
+            except ValueError:
+                bics[n - 1] = [np.inf, n]
+            else:
+                bics[n - 1] = [gmm.bic(sample_data.reshape(-1, 1)), n]
+        min_bic_index = np.argmin(bics[:, 0])
+        best_n = int(bics[min_bic_index, 1])
+
+        # With less components, we need to be more conservative
+        # Hence the standard_deviation multiplier is higher
+        std_dev_multi_dict = {
+            1: (2.0, 0.5),
+            2: (2.2, 0.65),
+            3: (3.0, 0.8),
+        }
+
+        # Redo the best with more iterations
+        self.gmm_estimator = GaussianMixture(
+            n_components=best_n,
+            covariance_type=covariance_type,
+            max_iter=300,
+            random_state=42,
+            init_params="k-means++",
+        )
+        self.gmm_estimator.fit(sample_data.reshape(-1, 1))
+
+        # Extract the means and variances
+        means = self.gmm_estimator.means_.flatten()
+        covariances = self.gmm_estimator.covariances_  # (n_components, n_features, n_features)
+        variances = covariances.flatten()  # n_features is 1, so this is n_components
+
+        # Pick the GMM component which is closest to the mean of the volume
+        volume_mean = np.mean(sample_data)
+        closest_mean_index = np.argmin(np.abs(means - volume_mean))
+        mean_to_use = means[closest_mean_index]
+        std_to_use = np.sqrt(variances[closest_mean_index])
+
+        low_variance_mult, high_variance_mult = std_dev_multi_dict[best_n]
+
+        low_limit, high_limit = (
+            mean_to_use - low_variance_mult * std_to_use,
+            mean_to_use + high_variance_mult * std_to_use,
+        )
+        # Ensure that the limits are within the range of the volume
+        low_limit = float(max(low_limit, np.min(sample_data)))
+        high_limit = float(min(high_limit, np.max(sample_data)))
+        return low_limit, high_limit
+
+    def _objective_function(self, params):
+        return self.compute_contrast_limit(
+            params["low_variance_mult"],
+            params["high_variance_mult"],
+        )
+
+    def _define_parameter_space(self, parameter_optimizer):
+        """NOTE: the range here is very small, for real-tuning, it should be larger."""
+        parameter_optimizer.space_creator(
+            {
+                "low_variance_mult": {"type": "uniform", "args": [2.2, 2.21]},
+                "high_variance_mult": {"type": "uniform", "args": [0.6, 0.61]},
+            },
+        )
+
+
+class CDFContrastLimitCalculator(ContrastLimitCalculator):
+
+    def __init__(self, volume: Optional["np.ndarray"] = None):
+        """Initialize the contrast limit calculator.
+
+        Parameters
+        ----------
+            volume: np.ndarray or None, optional.
+                Input volume for calculating contrast limits.
+        """
+        super().__init__(volume)
+        self.cdf = None
```

Comment:
Same here, seeing as `self.cdf` is used, perhaps it should be set as a simple variable where it's used and the `__init__` method should be removed.